skema/text_reading/text_reading/src/main/scala/org/ml4ai/skema/text_reading/CosmosTextReadingPipeline.scala

```diff
@@ -0,0 +1,157 @@
+      val groundedArguments =
+        e.arguments.mapValues(_.map(groundMention))
+
+
+      e.copy(arguments = groundedArguments)
+    }
+    // This is duplicated while we fix the Mention trait to define the abstract method copy
+    case e: RelationMention => {
+      val groundedArguments =
+        e.arguments.mapValues(_.map(groundMention))
+
+      e.copy(arguments = groundedArguments)
+    }
+    case m => m
+  }
+
+  logger.info("Initializing the OdinEngine ...")
+
+  // Read the configuration from the files
+  val generalConfig: Config = ConfigFactory.load()
+  val readerType: String = generalConfig[String]("ReaderType")
+  val defaultConfig: Config = generalConfig[Config](readerType)
+  val config: Config = defaultConfig.withValue("preprocessorType", ConfigValueFactory.fromAnyRef("PassThrough"))
+  val groundingConfig = generalConfig.getConfig("Grounding")
+
+  // Grounding parameters
+  private val ontologyFilePath = groundingConfig.getString("ontologyPath")
+  private val groundingAssignmentThreshold = groundingConfig.getDouble("assignmentThreshold")
+  private val grounder = MiraEmbeddingsGrounder(ontologyFilePath, None, lambda = 10, alpha = 1.0f) // TODO: Fix this @Enrique
+
+  // Odin Engine instantiation
+  private val odinEngine = OdinEngine.fromConfig(config)
+
+  // PDF converted to fix pdf tokenization artifacts
+  val pdfConverter = new TextConverter()
+  val languageModel = GigawordLanguageModel()
+  val preprocessors = Array(
+    new LinePreprocessor(),
+    new ParagraphPreprocessor(),
+    new UnicodePreprocessor(),
+    new CasePreprocessor(CasePreprocessor.defaultCutoff),
+    new NumberPreprocessor(NumberPreprocessor.Hyperparameters()),
+    new LigaturePreprocessor(languageModel),
+    new LineBreakPreprocessor(languageModel),
+    new WordBreakByHyphenPreprocessor(),
+    new WordBreakBySpacePreprocessor(languageModel) // This is by default NeverLanguageModel.
+  )
+  val pdf2txt = new Pdf2txt(pdfConverter, preprocessors)
+  ////
+
+  // cosmos stores information about each block on each pdf page
+  // for each block, we load the text (content) and the location of the text (page_num and block order/index on the page)
+  val loader = new CosmosJsonDataLoader
+
+  /**
+    * Runs the textReadingPipeline over a cosmos file
+    * @param jsonPath Path to the json file to annotate
+    * @return Mentions extracted by the TR textReadingPipeline
+    */
+  def apply(jsonPath: String): Seq[Mention] = {
+
+    // TODO: Make this interpretable
+    val textsAndLocations = loader.loadFile(jsonPath)
+    val textsAndFilenames = textsAndLocations.map(_.split("<::>").slice(0, 2).mkString("<::>"))
+    val locations = textsAndLocations.map(_.split("<::>").takeRight(2).mkString("<::>")) //location = pageNum::blockIdx
+
+    logger.info("started extracting")
+    // extract mentions form each text block
+    val mentions = for (tf <- textsAndFilenames) yield {
+      val Array(rawText, filename) = tf.split("<::>")
```

Comment:
So these were split, then put back together, and then split again?  I'd avoid that:

```scala
    val lines = loader.loadFile(jsonPath)
    val records: Seq[(String, String, Array[Int], Array[Int])] = lines.map { line =>
      val fields = line.split("<::>")
      val text = fields(0)
      val filename = fields(1)
      val pageNumbers = fields(2).split(",").map(_.toInt) // index may not be correct
      val blockIndexes = fields(3).split(",").map(_.toInt) // ditto

      (text, filename, pageNumbers, blockIndexes)
    }

    val mentions = for ((rawText, filename, _, _) <- records) yield {
```
and later
```scala
      val pageNum = records(id)._3
      val blockIdx = records(id)._4
```
Even so, a `Record` class would be preferable to numbers in the tuple.  Rather than having the `ArrayBuffer` soon thereafter, `flatMap()` could be used.

Reply:
Maybe even
```scala
      val Array(pageNumbers, blockIndexes) = fields.takeRight(2).map(_.split(",").map(_.toInt))
```

Reply:
Or
```scala
      val Array(text, filename, rest @ _*) = line.split("<::>")
      val Array(pageNumbers, blockIndexes) = rest.takeRight(2).map(_.split(",").map(_.toInt)).toArray

      (text, filename, pageNumbers, blockIndexes)
```